processout-sdk/src/main/java/com/processout/processout_sdk/ProcessOut.java

```diff
@@ -233,20 +242,103 @@ public void error() {
+                        fingerPrintWebView.removeAllViews();
+                        fingerPrintWebView.clearHistory();
+                        fingerPrintWebView.clearCache(true);
+                        fingerPrintWebView.onPause();
+                        // Removing the webview and returning the error
+                        ((ViewGroup) fingerPrintWebView.getParent()).removeView(fingerPrintWebView);
+                        fingerPrintWebView.destroy();
+                        handler.onError(new ProcessOutNetworkException("Fingerprint fallback timed out."));
+                    }
+                };
+
+                // Catch webview URL redirects
+                fingerPrintWebView.setWebViewClient(new WebViewClient() {
+                    @Override
+                    public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
+                        // Check if the current Android version is supported
+                        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.LOLLIPOP) {
+                            String token = request.getUrl().getQueryParameter("token");
+                            if (token != null) {
+                                // The webview successfully received the gateway token
+                                // We cancel the timeout handler
+                                timeOutHandler.removeCallbacks(timeoutClearer);
+                                makeCardPayment(invoiceId, request.getUrl().getQueryParameter("token"), handler);
+                            }
+                        } else {
+                            // Android version not supported for fingerprinting
+                            // We cancel the timeout handler
+                            timeOutHandler.removeCallbacks(timeoutClearer);
+                            handler.onError(new ProcessOutException("Unsupported Android version"));
+                        }
+                        return super.shouldOverrideUrlLoading(view, request);
+                    }
+                });
+
+                // Start the timeout
+                timeOutHandler.postDelayed(timeoutClearer, 10000);
+
+                // Load the fingerprint URl
+                fingerPrintWebView.loadUrl(cA.getValue());
+
+                // Add the webview to content
+                Activity a = (Activity) this.context;
+                FrameLayout rootLayout = a.findViewById(android.R.id.content);
+                rootLayout.addView(fingerPrintWebView);
+                break;
             default:
-                handler.onError(new ProcessOutAuthException("Unhandled three D S action:" + cA.getType().name()));
+                handler.onError(new ProcessOutException("Unhandled three D S action:" + cA.getType().name()));
                 break;
         }
     }
 
 
+    public void continueThreeDSVerification(final String invoiceId, String token, final ThreeDSVerificationCallback callback) {
```

Comment:
What does this function do? I can't see it used anywhere

Reply:
It is used when coming back to the app here:
https://github.com/processout/processout-android/pull/8/files#diff-6b7d01fd02a31e2be5a308d336f14f91R47